metricbeat/mb/mb.go

```diff
@@ -94,6 +104,44 @@ func (m *BaseModule) UnpackConfig(to interface{}) error {
 	return m.rawConfig.Unpack(to)
 }
 
+// ReConfigure re-configures the module with the given raw configuration
+func (m *BaseModule) ReConfigure(config *common.Config, register *Register) error {
+	currName := m.name
+	currConfig := m.config
+	currRawConfig := m.rawConfig
+
+	var chkConfig struct {
+		Module string `config:"module"`
+	}
+	if err := config.Unpack(&chkConfig); err != nil {
+		return errors.Wrap(err, "error parsing new module configuration")
+	}
+
+	// Don't allow module name change
+	if chkConfig.Module != "" && chkConfig.Module != currName {
+		return fmt.Errorf("cannot change module name from %v to %v", currName, chkConfig.Module)
+	}
+
+	if err := config.SetString("module", -1, currName); err != nil {
+		return errors.Wrap(err, "unable to set existing module name in new configuration")
+	}
+
+	if err := config.Unpack(&m.config); err != nil {
+		return errors.Wrap(err, "error parsing new module configuration")
+	}
+	m.rawConfig = config
+
+	// Calling initMetricSets ensures proper validation checks
+	if _, err := initMetricSets(register, m); err != nil {
```

Comment:
Don't worry about the `Close()` part, I think this shouldn't be needed if we have only called to the metricset factory, I think that all modules that implement `Close()` instantiate the things that need cleanup on `Fetch()`.

Reply:
Gotcha, thanks for the explanation. Indeed the only reason I'm doing this here is to prevent someone from passing in an invalid metricset in the new config. And the only way to validate that is against the register so that has to be passed in to this method (which, I agree, is ugly).

Perhaps I will just document in this method's godoc that if the new config contains metricsets, they need to be valid?

Reply:
Okay, I've removed `register` in 6d13e2d56 and added a godoc on top of this method explaining the precondition it expects. Let me know what you think of this.